File: C# - Simple Updater Library/Downloader.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Net;
using System.Threading;

namespace Simple_Updater_Library
{
    partial class SimpleUpdater
    {
        // Private vars for the Downloader
        private Queue<string> file_to_download;
        private long bytesdownloaded;
        private long totalbytestoreceive;
        private bool completed;

        private void DownloadFiles_Downloader()
        {
            // Initialize vars
            completed = false;
            bytesdownloaded = 0;
            file_to_download = new Queue<string>();

            foreach (KeyValuePair<string, File> entry in this.server_files)
            {
                // Create the directory
                try
                {
                    string dest_path_file = Path.Combine(this.installation_path, entry.Value.filename);
                    string new_dir_path = Path.GetDirectoryName(dest_path_file);
                    Directory.CreateDirectory(new_dir_path);
                }
                catch
                {
                    // Already exists
                }

                // Enqueue all files to the queue and all files will be processed one by one 
                this.file_to_download.Enqueue(entry.Value.filename);
            }

            // Don't block the main thread => Async thread
            Status_Changed(4);
            Thread download = new Thread(() => DownloadFile());
            download.Start();
        }

        private void DownloadFile()
        {
            // Check if there is any file in the queue
            if (this.file_to_download.Any())
            {
                WebClient client = new WebClient();

                // Set events for the client downloader
                client.DownloadProgressChanged += client_DownloadProgressChanged;
                client.DownloadFileCompleted += client_DownloadFileCompleted;

                string filename = file_to_download.Dequeue();

                Uri uri = new Uri(this.server_url + "/files/" + filename);
                string dest_path_file = Path.Combine(this.installation_path, filename);

                client.DownloadFileAsync(uri, dest_path_file);
            }
            else
            {
                // Download finished
                Download_Finished();
            }
        }

        private void client_DownloadFileCompleted(object sender, AsyncCompletedEventArgs e)
        {
            if (e.Error != null)
            {
                Status_Changed(6);
                throw e.Error;
            }

            this.bytesdownloaded += this.totalbytestoreceive;

            this.totalbytestoreceive = 0;
            this.completed = false;

            Status_Changed(5);

            DownloadFile();
        }

        private void client_DownloadProgressChanged(object sender, DownloadProgressChangedEventArgs e)
        {
            this.totalbytestoreceive = e.TotalBytesToReceive;

            // Weird fix, this event is triggered two times for 100% - So check with 'completed' to avoid duplicates additions
            if (!this.completed && e.ProgressPercentage == 100)
            {
                this.completed = true;
                DownloadProgress_Changed(this.bytesdownloaded + e.BytesReceived);
            }
            else if (!this.completed)
            {
                DownloadProgress_Changed(this.bytesdownloaded + e.BytesReceived);
            }
        }
    }
}

File: JAVA - Simple Updater Library/src/fr/s0me1ne/simpleupdater/Downloader.java
package fr.s0me1ne.simpleupdater;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.FileOutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.file.Path;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;

class Downloader {

	String installation_path;
	String server_url;
	List<SimpleUpdaterEvents> listeners;
	
	 // Private vars for the Downloader
    private Queue<String> file_to_download;
    private long bytesdownloaded;
    private long numberOfBytesToDownload;
    Map<String, File> server_files;
	
    public Downloader(String installation_path, String server_url, Map<String, File> server_files, long numberOfBytesToDownload, List<SimpleUpdaterEvents> listeners)
    {
    	this.server_url = server_url;
    	this.installation_path = installation_path;
    	this.listeners = listeners;
    	this.server_files = server_files;
    	this.numberOfBytesToDownload = numberOfBytesToDownload;
    }
   
    void DownloadFiles_Downloader()
    {   	
        // Initialize vars
        bytesdownloaded = 0;
        file_to_download = new LinkedList<String>();

        for(Map.Entry<String, File> entry : this.server_files.entrySet())
        {
            // Create the directory
            try
            {
            	Path dest_path_file = java.nio.file.Paths.get(this.installation_path, entry.getValue().filename);
            	new java.io.File(dest_path_file.getParent().toString()).mkdirs();
            }
            catch (Exception e)
            {
                // Already exists
            }

            // Enqueue all files to the queue and all files will be processed one by one 
            this.file_to_download.add(entry.getValue().filename);
        }

        // Don't block the main thread => Async thread
        //Status_Changed(4);
        Thread download = new Thread() {
			@Override
			public void run()
			{
				try 
				{
					ThrowEvent_StatusCodeChanged(4);
					DownloadFile();
				} 
				catch (Exception e) 
				{
					ThrowEvent_StatusCodeChanged(6);
					e.printStackTrace();
				}
			}
        };
        download.start();
    }
    
    private Runnable DownloadProgress_Changed(long bytesreceived, long numberOfBytesToDownload)
    {
    	float percent;
        // Division by 0 ? Don't know if it can occur
        try
        {
            percent = (float)(bytesreceived) / (float)numberOfBytesToDownload;
        }
        catch (Exception e)
        {
            percent = 100;
        }
        
        for(SimpleUpdaterEvents e : this.listeners)
		{
			e.OnDownloadProgressChanged(bytesreceived, percent*100);
		}

		return null;
    }
    
    private void DownloadFile() throws Exception
    {
        // Check if there is any file in the queue
        if (!this.file_to_download.isEmpty())
        {
        	String filename = file_to_download.poll();
        	
        	URL website = new URL(this.server_url + "/files/" + URLEncoder.encode(filename, "UTF-8").replaceAll("\\+", "%20").replaceAll("%5C", "/"));
        	
        	HttpURLConnection httpConnection = (HttpURLConnection) (website.openConnection());
        	//this.totalbytestoreceive = httpConnection.getContentLength();

            BufferedInputStream in = new BufferedInputStream(httpConnection.getInputStream());
            FileOutputStream fos = new FileOutputStream(java.nio.file.Paths.get(this.installation_path, filename).toString());
            BufferedOutputStream bout = new BufferedOutputStream(fos, 2048);
            
            byte[] data = new byte[2048];
            int x = 0;
            while ((x = in.read(data, 0, 2048)) >= 0) 
            {
            	this.bytesdownloaded += x;

            	// Create new thread to trigger event download progress
            	new Thread(DownloadProgress_Changed(this.bytesdownloaded, this.numberOfBytesToDownload)).start();
            	
            	bout.write(data, 0, x);
            }
            bout.close();
            in.close();
            ThrowEvent_StatusCodeChanged(5);
            DownloadFile();
        }
        else
        {
            // Download finished
            Download_Finished();
        }
    }
    
    private void Download_Finished()
    {
    	for(SimpleUpdaterEvents e : this.listeners)
		{
			e.OnDownloadFinished(this.bytesdownloaded);
			e.OnStatusChanged(8);
			e.OnStatusChanged(0);
		}
    }
    
	private void ThrowEvent_StatusCodeChanged(int code)
	{
		for(SimpleUpdaterEvents e : this.listeners)
		{
			e.OnStatusChanged(code);
		}

	}


}
